src/Microsoft.DotNet.GenAPI/Shared/CSharpBuilder.cs

```diff
@@ -0,0 +1,228 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.DotNet.GenAPI.Shared;
+
+/// <summary>
+/// Processes assemly symbols to build correspoding structures in C# language.
+/// Depend on _syntaxWriter impelemtenation, the result could be C# source file, xml etc.
+/// </summary>
+public class CSharpBuilder : AssemblySymbolTraverser, IAssemblySymbolWriter, IDisposable
+{
+    private readonly ISyntaxWriter _syntaxWriter;
+
+    public CSharpBuilder(IAssemblySymbolOrderProvider orderProvider,
+        IAssemblySymbolFilter filter, ISyntaxWriter syntaxWriter)
+        : base(orderProvider, filter) => _syntaxWriter = syntaxWriter;
+
+    public void WriteAssembly(IAssemblySymbol assembly) => Visit(assembly);
+
+    protected override IDisposable ProcessBlock(INamespaceSymbol namespaceSymbol)
+    {
+        var namespacePath = new List<string>();
+
+        foreach (var part in namespaceSymbol.ToDisplayParts(
+            AssemblySymbolDisplayFormats.NamespaceDisplayFormat))
+        {
+            if (part.Kind == SymbolDisplayPartKind.NamespaceName)
+            {
+                namespacePath.Add(part.ToString());
+            }
+        }
+        return _syntaxWriter.WriteNamespace(namespacePath);
+    }
+
+    protected override IDisposable ProcessBlock(INamedTypeSymbol namedType)
+    {
+        var typeName = namedType.ToDisplayString(AssemblySymbolDisplayFormats.NamedTypeDisplayFormat);
+
+        var accessibility = BuildAccessibility(namedType as ISymbol);
+        var keywords = GetKeywords(namedType as ITypeSymbol);
```

Comment:
`as` returns `null` silently and `GetKeywords` doesn't seem to accept a `null` correctly. I'd rather do this:
```suggestion
        var keywords = GetKeywords((ITypeSymbol)namedType);
```

This will throw a cast exception instead and make it more clear on what is broken if this ever happens